db/migrate/20240109180000_fix_unread_comments_inconsistencies.rb

```diff
@@ -0,0 +1,54 @@
+# Fixes the unread_comments flag for all users. Unintended behavior was
+# introduced in pull request #515. Migration introduced in #587.
+# Behavior fixed in #585.
+#
+# This migration is generally *not* idempotent since users might have interacted
+# with the website since the migration was run and thus they will probably have
+# different unread comments flags as the ones at the time of the migration.
+#
```

Comment:
When I did the migration in my development setting that had no faulty unread comments flags, I obtained what I expected:

```
Ran through 11 users (unread comments flag)
Fixed 0 users (unread comments flag)
```

Directly after that (without any user interactions in the meantime) I rolled back and obtained:

```
Ran through 11 users (unread comments flag)
Fixed 2 users (unread comments flag)
```

That makes me a little suspicious in several ways:

- Since we only have a `change` method, what does Rails do when rolling back? I would assume it executes the `change` method again, since there is no way for it to deduce how to revert the migration (in contrast to the usual `add_column` migrations). The SQL statements in the log seem to confirm this. Would it not be better to put the content of the current `change` method into the `up` method and define an empty `down` method?
- If in the above scenario the change method in the migration did not touch the database and no user interactions happened, how can an application of the same method suddenly result in 2 changed user records? Also, I see nothing in the SQL logs in the down migration that indicates any actual updates. Maybe there is something wrong with the return value in `fix_unread_comments_flag`?

Reply:
Mmm, I cannot reproduce this scenario in my local setup. Have you checked that entries in the `unread_comments` column actually change when it says "Fixed 2 users"? What command did you use to roll back?

With regards to the `change` method. You have a point here as I already marked the migration as *not* reversible in the comment. We should put the code in the `up` method and [throw an error to prevent reverts](https://guides.rubyonrails.org/active_record_migrations.html#throwing-an-error-to-prevent-reverts).

Reply:
I throw an error in 144141245e now.

Reply:
Oh, I think I found the flaw in my implementation and will fix it soon.

Reply:
Thanks for noting, there were some serious flaws in my implementation that I addressed in 97f5ac3. Foremost, I only checked the `Reader` for unread comments but forgot about the case where there is no Reader object, but where there is actually unread media present.

Reply:
Raising an `ActiveRecord::IrreversibleMigration` in the down migration is fine with me.